internal/plugin/exec/exec.go

```diff
@@ -0,0 +1,213 @@
+		Impl:    p.Impl,
+		Mappers: p.Mappers,
+		Logger:  p.Logger,
+	})
+	return nil
+}
+
+type ioWriter struct {
+	ctx    context.Context
+	stderr bool
+	client pb.ExecSessionServiceClient
+}
+
+func (i *ioWriter) Write(p []byte) (n int, err error) {
+	_, err = i.client.Output(i.ctx, &pb.ExecSession_OutputRequest{
+		Data:   p,
+		Stderr: i.stderr,
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return len(p), nil
+}
+
+func (p *ExecPlugin) GRPCClient(
+	ctx context.Context,
+	broker *plugin.GRPCBroker,
+	c *grpc.ClientConn,
+) (interface{}, error) {
+	client := pb.NewExecSessionServiceClient(c)
+
+	input, err := client.Input(ctx, &empty.Empty{})
+	if err != nil {
+		return nil, err
+	}
+
+	stdinR, stdinW := io.Pipe()
+
+	// 2 is picked randomly, just to try not to block.
+	wsUpdates := make(chan component.WindowSize, 2)
+
+	go func() {
+		for {
+			req, err := input.Recv()
+			if err != nil {
+				p.Logger.Debug("exec plugin input stream exitted", "error", err)
+				return
+			}
+
+			switch v := req.Input.(type) {
+			case *pb.ExecSession_InputRequest_Data:
+				p.Logger.Debug("exec plugin client, input received", "input", v.Data)
+
+				stdinW.Write(v.Data)
+			case *pb.ExecSession_InputRequest_WindowSize:
+				update := component.WindowSize{
+					Height: int(v.WindowSize.Height),
+					Width:  int(v.WindowSize.Width),
+				}
+				select {
+				case <-ctx.Done():
+					return
+				case wsUpdates <- update:
+					// ok
+				}
+			case *pb.ExecSession_InputRequest_InputClosed:
+				stdinW.Close()
+			}
+		}
+	}()
+
+	esi := &component.ExecSessionInfo{
+		Input: stdinR,
+		Output: &ioWriter{
+			ctx:    ctx,
+			client: client,
+		},
+		Error: &ioWriter{
+			ctx:    ctx,
+			stderr: true,
+			client: client,
+		},
+		WindowSizeUpdates: wsUpdates,
+	}
+
+	return esi, nil
+}
+
+// execServer is a gRPC server that the client talks to and calls a
+// real implementation of the component.
+type execServer struct {
+	Impl    *component.ExecSessionInfo
+	Mappers []*argmapper.Func
+	Logger  hclog.Logger
+}
+
+func (s *execServer) Output(
+	ctx context.Context,
+	req *pb.ExecSession_OutputRequest,
+) (*empty.Empty, error) {
+	var err error
+
+	if req.Stderr {
+		_, err = s.Impl.Error.Write(req.Data)
+	} else {
+		_, err = s.Impl.Output.Write(req.Data)
+	}
+
+	return &empty.Empty{}, err
+}
+
+func (s *execServer) Input(_ *empty.Empty, stream pb.ExecSessionService_InputServer) error {
+	s.Logger.Warn("starting exec server input")
+
+	readCh := make(chan []byte)
+
+	// rather than make this a chan struct{} and just close it to signal to the
+	// select loop, we send a bool and leave tha channel open. The reason being
+	// that the select loop will continue to run after input has closed, and if we
+	// close this channel, the select loop will fire constantly on this channel.
+	// So instead we just send a bool and orphan the channel, knowing that the select
+	// loop will stop when the stream context is done.
+	closedCh := make(chan bool)
```

Comment:
I'd buffer this 1 value so that `closedCh <- true` never blocks and dangles a goroutine below.